### PR TITLE
Extend user & group configuration parameters

### PR DESCRIPTION
**Pull request by mskarbe**
This PR introduces more configuration fields in order to enable `Gman` to manage them via API.
Group configuration was expanded by fields: Access Type settings, is archived.
User configuration was expanded by fields: aliases, phones & address, recovery phone & recovery email, employee & location information.
For the group's settings, the second API service client had to be added (GroupsSettings API client).

The `README` and example config yaml were updated accordingly to reflect new possibilities of `Gman`.

Moreover, the validation of the new fields was added. Also, the validation now collects the errors in the config file and followingly returns them all at once (i.e. does not abort after finding the first error).

Also, general error handling was improved (unwanted fatal() calls were removed and the failure messages were refined).

### CODE REVIEW

File: config.example.yaml [hunk 1]
```diff
@@ -34,7 +60,46 @@ users:
...
+    who_can_post: ALL_MEMBERS_CAN_POST
+    # Permission to join group. Possible values are:
+    # - ANYONE_CAN_JOIN: Anyone in the account domain 
+    # - ALL_IN_DOMAIN_CAN_JOIN: Any Internet user who can access your Google Groups service 
+    # - INVITED_CAN_JOIN: Candidates for membership can be invited to join
+    # - CAN_REQUEST_TO_JOIN: Non members can request an invitation to join
+    who_can_join: CAN_REQUEST_TO_JOIN
+    # Identifies whether members external to your organization can join the group. 
+    # Possible values are: "true" or "false"
+    allow_external_members: "false"
```
Comment (xrstf) [inline config.example.yaml:99]: Why are booleans strings in the YAML?

Comment (mskarbe) [reply config.example.yaml:99]: Because google API has those fields natively as a string and I have just aligned to it. 
I can add type conversions for those fields!

File: pkg/config/config.go [hunk 2]
```diff
@@ -76,57 +105,128 @@ func SaveToFile(config *Config, filename string) error {
 	return nil
 }
 
-func (c *Config) Validate() error {
+func (c *Config) Validate() []error {
+	var allTheErrors []error
+	re := regexp.MustCompile("^[a-zA-Z0-9.!#$%&'*+/=?^_`{|}~-]+@[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?(?:\\.[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?)*$")
```
Comment (xrstf) [inline pkg/config/config.go:110]: `re` is not the best vriable name for what I *think* is an email syntax regex.

Validating email addresses is like the holy war. On one hand you have people who say

> Don't do a syntax validation at all. Just look for an @ sign and try to open an SMTP connection to the target host.

and then there are people who say

> basic validation (max length, must contain @, ...) is good, but don't go overboard.

and others that say

> I once found this gigantic 3-page regex that supposedly covers all regex, and now I have printed and framed it above my bed and use it everywhere. It says "works for all e-mail addresses" on the tin, so I trust it!

Over the years I started at the bottom and worked my way up. Nowadays I would only check for an @ character and maybe the max length. Because TBH, I would have to spend an hour just reading the regex if *anything* breaks. Meaning we would technically need unit tests for this monster.

You can also look for a ready-made Go module that already has an email validation.

TL;DR: I would not have such a complicated regex in my codebase without tests, or even at all.

File: pkg/config/config.go [hunk 3]
```diff
@@ -76,57 +105,128 @@ func SaveToFile(config *Config, filename string) error {
...
 	}
 
-	re := regexp.MustCompile("^[a-zA-Z0-9.!#$%&'*+/=?^_`{|}~-]+@[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?(?:\\.[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?)*$")
-
 	//validate users
 	userEmails := []string{}
 	for _, user := range c.Users {
 		if util.StringSliceContains(userEmails, user.PrimaryEmail) {
-			log.Fatalf("Validation failed: duplicate user defined (user: %s)", user.PrimaryEmail)
+			//return fmt.Errorf("duplicate user defined (user: %s)", user.PrimaryEmail)
```
Comment (xrstf) [inline pkg/config/config.go:123]: The commented-out return statements can be removed :)

File: pkg/config/config.go (hunk 2 above)
Comment (xrstf) [reply pkg/config/config.go:110]: All of this is *unless* the backend system (GSuite in this case) does not explicitly define a format for email addreses. Looking further down, it seems like GSuite is defining a format for telephone numbers, so it's fine and encouraged to validate the format.

File: pkg/config/config.go (hunk 3 above)
Comment (mskarbe) [reply pkg/config/config.go:123]: ooh, just leftovers, im sorry! thank you for pointing out

File: pkg/config/config.go [hunk 4]
```diff
@@ -76,57 +105,128 @@ func SaveToFile(config *Config, filename string) error {
...
+			allTheErrors = append(allTheErrors, fmt.Errorf("duplicate group email defined (%s)", group.Email))
 		}
 
 		if re.MatchString(group.Email) == false {
-			log.Fatalf("Validation failed: invalid group email (%s)", group.Email)
+			//return fmt.Errorf("invalid group email (%s)", group.Email)
+			allTheErrors = append(allTheErrors, fmt.Errorf("group email is not a valid email-address (%s)", group.Email))
+		}
+
+		if !(strings.EqualFold(group.WhoCanContactOwner, "ALL_IN_DOMAIN_CAN_CONTACT") || strings.EqualFold(group.WhoCanContactOwner, "ALL_MANAGERS_CAN_CONTACT") || strings.EqualFold(group.WhoCanContactOwner, "ALL_MEMBERS_CAN_CONTACT") || strings.EqualFold(group.WhoCanContactOwner, "ANYONE_CAN_CONTACT")) {
```
Comment (xrstf) [inline pkg/config/config.go:190]: Just a question: The validation is case-insensitive, so does GSuite also not care or do we later force the user-configured value to be uppercase?

File: pkg/config/config.go [hunk 5]
```diff
@@ -76,57 +105,128 @@ func SaveToFile(config *Config, filename string) error {
...
+		}
+
+		if !(strings.EqualFold(group.WhoCanJoin, "CAN_REQUEST_TO_JOIN") || strings.EqualFold(group.WhoCanJoin, "INVITED_CAN_JOIN") || strings.EqualFold(group.WhoCanJoin, "ALL_IN_DOMAIN_CAN_JOIN") || strings.EqualFold(group.WhoCanJoin, "ANYONE_CAN_JOIN")) {
+			//return fmt.Errorf("wrong value specified for 'who_can_contact_owner' field (group: %s)", group.Name)
+			allTheErrors = append(allTheErrors, fmt.Errorf("wrong value specified for 'who_can_contact_owner' field (group: %s)", group.Name))
+		}
+
+		if !(strings.EqualFold(group.AllowExternalMembers, "true") || strings.EqualFold(group.AllowExternalMembers, "false")) {
+			//return fmt.Errorf("wrong value specified for 'allow_external_members' field (group: %s)", group.Name)
+			allTheErrors = append(allTheErrors, fmt.Errorf("wrong value specified for 'allow_external_members' field (group: %s)", group.Name))
```
Comment (xrstf) [inline pkg/config/config.go:217]: All of this can disappear when we just specify the field as `bool` in our struct, instead of `string`.

File: pkg/config/config.go [hunk 6]
```diff
@@ -135,16 +235,24 @@ func (c *Config) Validate() error {
...
 		if util.StringSliceContains(ouNames, ou.Name) {
-			log.Fatalf("Validation failed: duplicate org unit defined (%s)", ou.Name)
+			//return fmt.Errorf("duplicate org unit defined (%s)", ou.Name)
+			allTheErrors = append(allTheErrors, fmt.Errorf("duplicate org unit defined (%s)", ou.Name))
 		}
 
 		if ou.ParentOrgUnitPath[0] != '/' {
-			log.Fatalf("Validation failed: wrong ParentOrgUnitPath specified for org unit (%s)", ou.Name)
+			//return fmt.Errorf("wrong ParentOrgUnitPath specified for org unit (%s)", ou.Name)
+			allTheErrors = append(allTheErrors, fmt.Errorf("wrong ParentOrgUnitPath specified for org unit (%s)", ou.Name))
```
Comment (xrstf) [inline pkg/config/config.go:244]: What does "wrong" mean? How about `ParentOrgUnitPath must start with a slash` as an error?

Also, does this work when the field is left empty entirely? Wouldn't this then panic because of the `[0]`?

File: pkg/glib/glib.go [hunk 7]
```diff
@@ -5,36 +5,61 @@ import (
...
 	if err != nil {
-		return nil, fmt.Errorf("ReadFile(clientSecretFile): %v", err)
+		return nil, fmt.Errorf("Unable to read json credentials (clientSecretFile): %v", err)
 	}
 
-	config, err := google.JWTConfigFromJSON(jsonCredentials, admin.AdminDirectoryUserScope, admin.AdminDirectoryGroupScope, admin.AdminDirectoryGroupMemberScope, admin.AdminDirectoryOrgunitScope)
+	config, err := google.JWTConfigFromJSON(jsonCredentials, admin.AdminDirectoryUserScope, admin.AdminDirectoryGroupScope, admin.AdminDirectoryGroupMemberScope, admin.AdminDirectoryOrgunitScope, admin.AdminDirectoryResourceCalendarScope)
 	if err != nil {
-		return nil, fmt.Errorf("JWTConfigFromJSON: %v", err)
+		return nil, fmt.Errorf("Unable to process credentials: %v", err)
```
Comment (xrstf) [inline pkg/glib/glib.go:29]: error strings should always start with a lowercase letter, because errors are often concatenated like "something failed: this did not work: omg what?: file does not exist`. Exception is of course if the first word is a special noun or initials (like `LoadBalancer could not be reconciled`).

Comment (xrstf) [reply pkg/glib/glib.go:29]: See https://github.com/golang/go/wiki/CodeReviewComments#error-strings for more information.

File: pkg/glib/glib.go [hunk 8]
```diff
@@ -131,9 +244,121 @@ func createGSuiteUserFromConfig(user *config.UserConfig) *admin.User {
...
+		}
+	}
+
+	if googleUser.Organizations != nil {
+		for _, org := range googleUser.Organizations.([]interface{}) {
+			configUser.Employee.Department = fmt.Sprint(org.(map[string]interface{})["department"])
+			configUser.Employee.JobTitle = fmt.Sprint(org.(map[string]interface{})["title"])
+			configUser.Employee.Type = fmt.Sprint(org.(map[string]interface{})["description"])
+			configUser.Employee.CostCenter = fmt.Sprint(org.(map[string]interface{})["costCenter"])
+		}
```
Comment (xrstf) [inline pkg/glib/glib.go:333]: In the future we maybe should find a way to write this in a saner way. These many inline type assertions are a big code smell and I cannot understand why the GSuite SDK thingy is so unspecific/unhelpful.

Nothing that needs to be addressed in this PR though.

File: pkg/sync/sync.go [hunk 9]
```diff
@@ -85,22 +84,34 @@ func SyncUsers(ctx context.Context, clientService *admin.Service, cfg *config.Co
...
 		if usersToDelete != nil {
 			log.Println("✁ Deleting...")
 			for _, user := range usersToDelete {
-				glib.DeleteUser(*clientService, user)
-				log.Printf("\t- user: %s\n", user.PrimaryEmail)
+				err := glib.DeleteUser(*clientService, user)
+				if err != nil {
+					return fmt.Errorf("⚠ Failed to delete user %s: %v.", user.PrimaryEmail, err)
+				} else {
+					log.Printf("\t- user: %s\n", user.PrimaryEmail)
```
Comment (xrstf) [inline pkg/sync/sync.go:102]: I would discourage tabs here, if you aligned your text to match the emoji-width (for the warning sign icons and stuff). If you just want a random indentation it's fine, otherwise my tabwidth is most likely not your tabwidth and the output can look weird for me.

File: config.example.yaml (hunk 1 above)
Comment (xrstf) [reply config.example.yaml:99]: I think that would be good. We (I think) don't follow the GSuite data structures to the letter in our config.yaml, so in this case I would support defining a "sane" YAML and then converting the bool to a string (`fmt.Sprintf("%v", val)`) when talking to the API. :-)

File: pkg/config/config.go (hunk 4 above)
Comment (mskarbe) [reply pkg/config/config.go]: Indeed, it is case sensitive. My mistake. Changed to require upper-case values. 
I hope this fixes it: https://github.com/kubermatic-labs/gman/pull/11/commits/4b47f836cb50d908839f5ffef246927836676bce#diff-091ace02b49a46f55f2586cfec89a160R195

File: pkg/config/config.go (hunk 2 above)
Comment (mskarbe) [reply pkg/config/config.go]: Very good points, thank you!
I have changed the address validation to check for the presence of @ and the length: https://github.com/kubermatic-labs/gman/pull/11/commits/4b47f836cb50d908839f5ffef246927836676bce#diff-091ace02b49a46f55f2586cfec89a160R260 
I hope it is better now.

File: config.example.yaml (hunk 1 above)
Comment (mskarbe) [reply config.example.yaml]: I have introduced the conversion with `ParseBool` (and `FormatBool` for other way around), I hope it is okay. 
https://github.com/kubermatic-labs/gman/pull/11/commits/4b47f836cb50d908839f5ffef246927836676bce#diff-65be9b04993ff3a5c2435628e9035455R455

File: pkg/config/config.go (hunk 6 above)
Comment (mskarbe) [reply pkg/config/config.go]: True, resulted in panic. I have added to check at first for the presence of the value: https://github.com/kubermatic-labs/gman/pull/11/commits/4b47f836cb50d908839f5ffef246927836676bce#diff-091ace02b49a46f55f2586cfec89a160R235

**Review by xrstf — APPROVED** [review]
/approve
